homeassistant/components/upnp/device.py

```diff
@@ -154,3 +157,18 @@ async def async_get_traffic_data(self) -> Mapping[str, Any]:
             PACKETS_RECEIVED: values[2],
             PACKETS_SENT: values[3],
         }
+
+    async def async_get_status(self) -> Mapping[str, Any]:
+        """Get connection status, uptime, and external IP."""
+        _LOGGER.debug("Getting status for  device: %s", self)
+
+        values = await asyncio.gather(
+            self._igd_device.async_get_status_info(),
+            self._igd_device.async_get_external_ip_address(),
+        )
+
+        return {
```

Comment:
Is it possible that data won't contain the keys `WANIP` and/or `UPTIME`? If so, this will error. Both the [IGDv1](http://www.upnp.org/specs/gw/UPnP-gw-WANIPConnection-v1-Service.pdf) and [IGDv2](http://www.upnp.org/specs/gw/UPnP-gw-WANIPConnection-v2-Service.pdf) specifications say it is mandatory, but not all devices implement it properly.

Reply:
Also, `self._igd_device.async_get_status_info()` can return `None`, in case no UPnP-action is provided by the device. (Should be, as it is required by the specs, but not all devices follow the specs completely.) Same for `self._igd_device.async_get_external_ip_address()`.

The `DataUpdateCoordinator` will catch the error, but I think it should be handled properly.

Reply:
Added coding for WANSTATUS and UPTIME within device.py to return None if the result from self._igd_device.async_get_status_info() is None. For self._igd_device.async_get_external_ip_address() not an issue as it would just return None then.

Within the attributes I now check and only set the attribute if it's value is not None.
Further, if WANSTATUS is not None then I make the binary sensor unavailable.